internal/contentmanager/dynamicsource.go

```diff
@@ -0,0 +1,183 @@
+package contentmanager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/informers"
+	cgocache "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// DynamicInformerSource is an implementation of the
+// ReaderCloserSyncingSource interface. It is used
+// to create an informer, using the provided dynamic informer
+// factory, for the configured GroupVersionResource.
+//
+// The informer is configured with a WatchEventErrorHandler that
+// stops the informer, and if it had previously synced successfully
+// it attempts to requeue the provided Owner for reconciliation and
+// calls the provided OnWatchError function.
+type DynamicInformerSource struct {
+	// DynamicInformerFactory is the dynamicinformer.DynamicSharedInformerFactory
+	// that is used to generate the informers configured on this sources startup
+	DynamicInformerFactory dynamicinformer.DynamicSharedInformerFactory
+	// GVR is the GroupVersionResource that this source is responsible
+	// for creating and configuring an informer for
+	GVR schema.GroupVersionResource
+	// Owner is the client.Object that owns the managed content that this
+	// source will be creating an informer to react to events for. This
+	// field is used to attempt to requeue the owning client.Object for
+	// reconciliation on a watch error after a previously successful sync
+	Owner client.Object
+	// Handler is the handler.EventHandler that is used to react to events
+	// received by the configured informer
+	Handler handler.EventHandler
+	// Predicates are the predicate.Predicate functions used to determine
+	// if a triggered event should be reacted to
+	Predicates []predicate.Predicate
+	// OnWatchError is the callback function that is used when the informer
+	// initially synced successfully and later encountered an error
+	OnWatchError func()
+	errOnce      sync.Once
+	errChan      chan error
+	syncedChan   chan struct{}
+	mu           sync.RWMutex
+	synced       bool
+	cancel       context.CancelFunc
+	ctx          context.Context
+	informer     informers.GenericInformer
+}
+
+var _ ReaderCloserSyncingSource = (*DynamicInformerSource)(nil)
+
+func (dis *DynamicInformerSource) String() string {
+	return "contentmanager.DynamicInformerSource"
```

Comment:
We should include more context as to the data used in this source. Thinking along the lines of GVR, Owner Type/name, etc.